File: python3/min_cost_to_connect_all_points.py
from typing import List

"""
Space : O(n)
Time  : O(n^2)

Method:
Minimum Spanning Tree
"""

class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        d = {(x, y): float('inf') if i else 0 for i,
             (x, y) in enumerate(points)}

        ans = 0
        while d:
            x1, y1 = min(d, key=d.get)
            ans += d.pop((x1, y1))

            for x2, y2 in d:
                d[(x2, y2)] = min(d[(x2, y2)], (abs(x1 - x2) + abs(y1 - y2)))

        return ans

File: golang/min_cost_to_connect_all_points.go
package main

import "math"

/*
Space	: O(n)
Time	: O(n^2)
*/

func minCostConnectPoints(points [][]int) int {
	type point struct {
		x int
		y int
	}

	ans := 0
	hashmap := map[point]int{}
	for i, item := range points {
		p := point{x: item[0], y: item[1]}
		if i == 0 {
			hashmap[p] = 0
		} else {
			hashmap[p] = math.MaxInt
		}
	}

	for len(hashmap) > 0 {
		var p point
		curVal := math.MaxInt

		for key, val := range hashmap {
			if val < curVal {
				p = key
				curVal = val
			}
		}
		delete(hashmap, p)
		ans += curVal

		for key, val := range hashmap {
			hashmap[key] = Min(val, Abs(p.x-key.x)+Abs(p.y-key.y))
		}
	}

	return ans
}
